docs/architecture/adr-059-evidence-composition-and-lifecycle.md

```diff
@@ -0,0 +1,304 @@
+  header *Header
+  commonHeight int64
+  vote *Vote
+
+  timestamp time.Time // once again taken from the block time at the height of the common header
+}
+```
+
+### Hybrid Framework
+
+We allow individual and batch evidence to coexist together, meaning that verification, broadcasting and committing is done depending on the evidence type. A hybrid model effectively undermines any possibility of generalization and Tendermint's evidence module will use switch statements and concrete types instead of interfaces to correctly process the evidence. One could argue that although we could see more evidence on the horizon, this won't be to the extent of justifying generalizing evidence and that each evidence can make it's own unique way through the pipeline and on to being committed to the chain and broadcasted to the application.
+
+A hybrid framework (as well as a batch framework) also surfaces the problem of the interface between Tendermint and the application. Currently, we have one concrete type that captures the relevant information needed to communicate to the application. This concrete type is formed per validator and therefore makes it more difficult to assess cartel (multi-validator) attacks. This begs the question of whether Tendermint should, at the application level, break down the evidence for each of the validators or alter the ABCI data structure to include multiple validators that were involved at that height and thus better capture the full extent of the misbehavior.
+
+Last important note is that we are closest to this implementation than any of the other two.
+
+#### An example of a possible implementation path
+
+We would have to change the interface as it's not possible for `LightClientAttackEvidence` to return a single `Address()` so instead we would have `Addresses()`. In addition the actual infringing validators aren't known in `LightClientAttackEvidence` until verification is done as we are taking the union between the validator set at the common height and at the conflicting height.
+
+```go
+type Evidence interface { //proposed
+  Height() int64                                     // height of the equivocation
+  Time() time.Time                                   // time of the equivocation
+  Addresses() []Address                              // addresses of the equivocating validators
+  Bytes() []byte                                     // bytes which comprise the evidence
+  Hash() []byte                                      // hash of the evidence
+  Equal(Evidence) bool                               // check equality of evidence
+
+  ValidateBasic() error
+  String() string
+}
+```
+
+Most likely we would therefore need to internally save the infringing validators which would make it easier to form abci evidence once the evidence has been committed on to chain. The `BlockExecutor` shouldn't have too much evidence specific logic in forming the abci evidence. Hence `LightClientAttackEvidence` should look like:
```

Comment:
At this point, we don't actually know which of the "infringing validators" have actually been malicious, right? We won't know that until the evidence has been verified?